### PR TITLE
[UP-413] Lab results document template example

### PR DESCRIPTION
**Pull request by 3ace**

### CODE REVIEW

File: templates/lab-results/pdf_lab_results.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+				if color, ok := levelColor[level]; ok {
+					return color
+				}
+
+				return "#000000"
+			},
+		},
+	}
+
+	// data := map[string]interface{}{
```
Comment (adrg) [inline templates/lab-results/pdf_lab_results.go:74]: Please remove the commented code if it's not needed.

File: templates/lab-results/templates/main.tpl [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+            </paragraph>
+        </table-cell>
+
+        <table-cell border-width-left="1" colspan="2">
+            <paragraph margin="0 10">
+                <text-chunk font="helvetica-bold">Physician Details</text-chunk>
+            </paragraph>
+        </table-cell>
+
+        <table-cell border-width-left="1">
```
Comment (adrg) [inline templates/lab-results/templates/main.tpl:37]: Seems like this `field name: field value` pattern repeats quite a lot in this section. Perhaps you can define a sub-template to show this pair of two table cells.

```xml
{{define "field"}}
    <table-cell border-width-left="1">
        <paragraph margin="0 10">
            <text-chunk font="helvetica-bold" color="#1B7A98">{{.FieldName}}</text-chunk>
        </paragraph>
    </table-cell>
    <table-cell>
        <paragraph margin="0 10 0 0">
            <text-chunk>{{.FieldValue}}</text-chunk>
        </paragraph>
    </table-cell>
{{end}}

{{template "patient-field" dict "FieldName" "DOB(y/m/d):" "FieldValue" $.Patient.Birthdate}}
```

If you make this template a bit more configurable (like the option to control the font size), maybe you can use it to show the fields above this section as well (`Specimen ID`, `Control ID`, `Acct #`, etc). That section looks similar as well.

File: templates/lab-results/templates/header.tpl [hunk 3]
```diff
@@ -0,0 +1,81 @@
+<division margin="10">
+    <table columns="2" column-width="0.5 0.5">
+        <table-cell border-width-bottom="2" border-color="#000000">
+            <image src="path('templates/res/logo.png')" margin="0 0 5 5"></image>
+        </table-cell>
+
+        <table-cell vertical-align="bottom" align="left" border-width-bottom="2" border-color="#000000">
```
Comment (adrg) [inline templates/lab-results/templates/header.tpl:7]: No need to specify the `align` attribute as it defaults to `left`.

File: templates/lab-results/pdf_lab_results.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+type Level string
+
+const (
+	Green  Level = "green"
+	Yellow       = "yellow"
+	Red          = "red"
+)
+
+var (
```
Comment (adrg) [inline templates/lab-results/pdf_lab_results.go:195]: Perhaps skip the `var` block as you are creating a single variable.

```go
var levelColor = map[Level]string{
	Green:  "#407505",
	Yellow: "#F5A623",
	Red:    "#FF0000",
}
```

File: templates/lab-results/pdf_lab_results.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	c := creator.New()
+	c.SetPageMargins(15, 15, 180, 100)
+
+	// Read main content template.
+	mainTpl, err := readTemplate("templates/main.tpl")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Read lab result data json.
```
Comment (adrg) [inline templates/lab-results/pdf_lab_results.go:38]: Can you change the comment to `Read lab result JSON data.` please?

File: templates/lab-results/pdf_lab_results.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+	// data := map[string]interface{}{
+	// 	"firmName":    "UniDoc Financial Firm",
+	// 	"firmAddress": "123 Main Street\nPortland, ME 12345\n(123) 456-789",
+	// }
+
+	if err := c.DrawTemplate(mainTpl, data, tplOpts); err != nil {
+		log.Fatal(err)
+	}
+
+	// Draw header.
```
Comment (adrg) [inline templates/lab-results/pdf_lab_results.go:83]: Can you change the comment to `Draw header and footer.` please?

File: templates/lab-results/pdf_lab_results.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
+package main
```
Comment (adrg) [inline templates/lab-results/pdf_lab_results.go:1]: Please add the usual file header examples have.

```go
/*
 * This example showcases the usage of creator templates by creating a sample
 * lab result document.
 *
 * Run as: go run pdf_lab_results.go
 */
```

**Review by adrg — COMMENTED** [review]
Looks good. I left a couple of minor observations.

**Review by kellemNegasi — COMMENTED** [review]
Looks good to me. 
I have left some comments.

**Comment by 3ace** [comment]
Hi @adrg @kellemNegasi Thanks for your inputs. I've updated this PR according to those inputs.

**Review by kellemNegasi — COMMENTED** [review]
Looks good to me.

File: templates/lab-results/README.md [hunk 8]
```diff
@@ -0,0 +1,9 @@
+Sample lab result document
```
Comment (adrg) [inline templates/lab-results/README.md:1]: I see that you are sometimes using `lab results`, sometimes `lab result` (singular) when you are referring to the example. Please choose one of them and stick with that. There are multiple occurrences of both throughout the code and `README.md` files.

File: templates/lab-results/templates/main.tpl [hunk 9]
```diff
@@ -0,0 +1,144 @@
+{{define "test-result-header"}}
+    <table-cell background-color="#1B7A98" align="center" vertical-align="middle">
+        <paragraph text-align="center">
+            <text-chunk font-size="10" color="#ffffff">{{ .Text }}</text-chunk>
+        </paragraph>
+    </table-cell>
+{{end}}
+
+{{define "test-result-row"}}
+    <table-cell align="center" vertical-align="top" border-width-bottom="1" border-color="#1B7A98">
```
Comment (adrg) [inline templates/lab-results/templates/main.tpl:10]: No need to specify the `vertical-align` attribute as `top` is the default value for that.

File: templates/lab-results/templates/main.tpl [hunk 10]
```diff
@@ -0,0 +1,144 @@
+{{define "test-result-header"}}
+    <table-cell background-color="#1B7A98" align="center" vertical-align="middle">
+        <paragraph text-align="center">
+            <text-chunk font-size="10" color="#ffffff">{{ .Text }}</text-chunk>
```
Comment (adrg) [inline templates/lab-results/templates/main.tpl:4]: No need to specify the `font-size` attribute in this case as `10` is the default value.

File: templates/lab-results/templates/header.tpl [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    </table-cell>
+    <table-cell>
+        <paragraph>
+            <text-chunk font-size="12">{{ .Value }}</text-chunk>
+        </paragraph>
+    </table-cell>
+{{end}}
+
+<division margin="10">
+    <table columns="2" column-width="0.5 0.5">
```
Comment (adrg) [inline templates/lab-results/templates/header.tpl:15]: No need to specify the `column-widths` attribute if all the columns need to be equal.

File: templates/lab-results/templates/header.tpl [hunk 12]
```diff
@@ -0,0 +1,45 @@
+{{define "info-field"}}
+    <table-cell>
+        <paragraph margin="0">
```
Comment (adrg) [inline templates/lab-results/templates/header.tpl:3]: No need to specify the `margin` attribute as it defaults to `0` anyway.

File: templates/lab-results/templates/main.tpl [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+        </paragraph>
+    </table-cell>
+{{end}}
+
+{{define "test-result-row"}}
+    <table-cell align="center" vertical-align="top" border-width-bottom="1" border-color="#1B7A98">
+        <paragraph text-align="center" line-height="1.1">
+            <text-chunk font-size="10">{{ .Text }}</text-chunk>
```
Comment (adrg) [inline templates/lab-results/templates/main.tpl:12]: No need to specify the `font-size` attribute in this case as 10 is the default value.

**Review by adrg — COMMENTED** [review]
Looks good. I left a couple of minor observations.

**Comment by 3ace** [comment]
@adrg I've updated this PR according to your input. Thanks

**Comment by gunnsth** [comment]
@3ace can you fix the conflict?

**Comment by 3ace** [comment]
@gunnsth resolved.

File: templates/lab-results/pdf_lab_results.go [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+	// Read lab results JSON data.
+	data, err := readResultData("lab_results.json")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Draw main content template.
+	tplOpts := &creator.TemplateOptions{
+		HelperFuncMap: template.FuncMap{
+			"extendDict": func(m map[string]interface{}, params ...interface{}) (map[string]interface{}, error) {
```
Comment (adrg) [inline templates/lab-results/pdf_lab_results.go:54]: Please remove the `extendDict` function as you don't seem to be using it.

**Review by adrg — COMMENTED** [review]
Looks good. I left one minor observation. Other than that, I think this example can be merged.

File: templates/lab-results/pdf_lab_results.go (hunk 14 above)
Comment (3ace) [reply templates/lab-results/pdf_lab_results.go]: removed on latest commit

**Review by adrg — APPROVED** [review]
Looks good. Thank you for addressing the review points.

**Pull request merged**